File: app/services/s3_service.py
import asyncio
import logging
from typing import Optional
import boto3
from botocore.exceptions import ClientError

from app.core.config import settings

logger = logging.getLogger(__name__)


class S3Service:
    def __init__(self):
        self.bucket_name = settings.AWS_STORAGE_BUCKET_NAME
        self.region = settings.AWS_S3_REGION_NAME

    def _get_client(self):
        return boto3.client(
            "s3",
            aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
            aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
            region_name=self.region,
        )

    def _upload_file_sync(
        self,
        file_bytes: bytes,
        object_name: str,
        content_type: str = "application/octet-stream",
    ) -> Optional[str]:
        """Synchronously upload bytes to AWS S3 bucket and return public URL."""
        if not settings.AWS_ACCESS_KEY_ID or not settings.AWS_SECRET_ACCESS_KEY:
            logger.warning("AWS S3 credentials missing. File upload skipped.")
            return None

        try:
            s3_client = self._get_client()
            s3_client.put_object(
                Bucket=self.bucket_name,
                Key=object_name,
                Body=file_bytes,
                ContentType=content_type,
            )
            file_url = f"https://{self.bucket_name}.s3.{self.region}.amazonaws.com/{object_name}"
            logger.info("Successfully uploaded file to S3: %s", file_url)
            return file_url
        except ClientError as e:
            logger.error("Failed to upload file to AWS S3: %s", e)
            return None

    async def upload_file(
        self,
        file_bytes: bytes,
        object_name: str,
        content_type: str = "application/octet-stream",
    ) -> Optional[str]:
        """Asynchronously upload file bytes to S3 without blocking the event loop."""
        return await asyncio.to_thread(
            self._upload_file_sync,
            file_bytes=file_bytes,
            object_name=object_name,
            content_type=content_type,
        )


s3_service = S3Service()

File: src/services/s3.service.ts
import path from 'path';
import fs from 'fs';
import { S3Client, PutObjectCommand, GetObjectCommand } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { env } from '../config/env';
import { logger } from '../config/logger';

class S3Service {
  private client: S3Client | null = null;

  constructor() {
    if (env.AWS_ACCESS_KEY_ID && env.AWS_SECRET_ACCESS_KEY && env.AWS_STORAGE_BUCKET_NAME) {
      this.client = new S3Client({
        region: env.AWS_S3_REGION_NAME,
        credentials: {
          accessKeyId: env.AWS_ACCESS_KEY_ID,
          secretAccessKey: env.AWS_SECRET_ACCESS_KEY,
        },
      });
    }
  }

  async uploadFile(
    fileBuffer: Buffer,
    fileName: string,
    contentType: string = 'application/octet-stream'
  ): Promise<string> {
    if (!this.client || !env.AWS_STORAGE_BUCKET_NAME) {
      return this.saveLocalFile(fileBuffer, fileName);
    }

    try {
      const cleanName = fileName.replace(/[^a-zA-Z0-9._-]/g, '_');
      const key = `uploads/${Date.now()}-${cleanName}`;
      const command = new PutObjectCommand({
        Bucket: env.AWS_STORAGE_BUCKET_NAME,
        Key: key,
        Body: fileBuffer,
        ContentType: contentType,
      });

      await this.client.send(command);
      return `https://${env.AWS_STORAGE_BUCKET_NAME}.s3.${env.AWS_S3_REGION_NAME}.amazonaws.com/${key}`;
    } catch (err: any) {
      logger.warn(`S3 upload error: ${err.message}. Falling back to local public storage.`);
      return this.saveLocalFile(fileBuffer, fileName);
    }
  }

  private saveLocalFile(fileBuffer: Buffer, fileName: string): string {
    const cleanName = fileName.replace(/[^a-zA-Z0-9._-]/g, '_');
    const safeName = `${Date.now()}-${cleanName}`;
    const uploadDir = path.resolve('uploads', 'products');
    if (!fs.existsSync(uploadDir)) {
      fs.mkdirSync(uploadDir, { recursive: true });
    }
    fs.writeFileSync(path.join(uploadDir, safeName), fileBuffer);
    logger.info(`Saved file to local public static directory: uploads/products/${safeName}`);
    return `/static/products/${safeName}`;
  }

  async getPresignedUrl(key: string, expiresInSeconds: number = 3600): Promise<string> {
    if (!this.client || !env.AWS_STORAGE_BUCKET_NAME) {
      return `/static/${key}`;
    }

    const command = new GetObjectCommand({
      Bucket: env.AWS_STORAGE_BUCKET_NAME,
      Key: key,
    });

    return getSignedUrl(this.client, command, { expiresIn: expiresInSeconds });
  }
}

export const s3Service = new S3Service();
